src/lib/parallel_scan/parallel_scan.ml

```diff
@@ -1,1328 +1,1367 @@
+        -> f_split:(b * b -> c -> c * c)
+        -> on_level:int
+        -> c t =
+     fun job_list tree ~weight_a ~weight_d ~f_split ~on_level ->
+      match tree with
+      | Node {depth; value; sub_tree} ->
+          if depth = on_level then job_list
+          else
+            let l, r = weight_a value in
+            let new_job_list = split job_list (f_split (l, r)) in
+            Double
+              (of_tree new_job_list sub_tree
+                 ~weight_a:(fun (a, b) -> (weight_a a, weight_a b))
+                 ~weight_d:(fun (a, b) -> (weight_d a, weight_d b))
+                 ~f_split:(fun ((x1, y1), (x2, y2)) (a, b) ->
+                   (f_split (x1, y1) a, f_split (x2, y2) b) )
+                 ~on_level)
+      | Leaf b ->
+          Double (split job_list (f_split (weight_d b)))
+
+    let of_list_and_tree lst tree on_level =
+      of_tree (Single lst) tree ~weight_a:fst
+        ~weight_d:(fun d -> (fst d, 0))
+        ~f_split:(fun (l, r) a -> (List.take a l, List.take (List.drop a l) r))
+        ~on_level
+
+    (*Just the nested data*)
+    let to_data : type a. a t -> a =
+     fun t ->
+      let rec go : type a. a t -> a * a =
+       fun data_list ->
+        match data_list with Single a -> (a, a) | Double js -> fst (go js)
       in
-      assert (total_blocks + 1 = Int.pow 2 x)
-    in
-    of_root_at_depth 0 ;
-    of_root_at_depth 1 ;
-    of_root_at_depth 2 ;
-    of_root_at_depth 3 ;
-    of_root_at_depth 4 ;
-    of_root_at_depth 5
-
-  (*Assuming that Base Somes are picked in the same order*)
-  let next_base_pos state cur_pos =
-    let p = parallelism state in
-    let next_pos = next_leaf_pos p cur_pos in
-    match Ring_buffer.read_i state.jobs next_pos with
-    | Base None ->
-        Some next_pos
-    | _ ->
-        None
+      fst @@ go t
+  end
 
-  let%test_unit "parallelism derived from jobs" =
-    let of_parallelism_log_2 x =
-      let s = create ~parallelism_log_2:x ~root_at_depth:1 in
-      assert (parallelism s = Int.pow 2 x)
-    in
-    of_parallelism_log_2 1 ;
-    of_parallelism_log_2 2 ;
-    of_parallelism_log_2 3 ;
-    of_parallelism_log_2 4 ;
-    of_parallelism_log_2 5 ;
-    of_parallelism_log_2 10
-
-  let visualize state ~draw_a ~draw_d =
-    let none = "_" in
-    let print s1 s2 = Printf.sprintf "(%s,%s)" s1 s2 in
-    let draw_job = function
-      | Job.Merge Empty ->
-          print none none
-      | Job.Merge (Lcomp x) ->
-          print (draw_a x) none
-      | Job.Merge (Rcomp y) ->
-          print none (draw_a y)
-      | Job.Merge (Bcomp (x, y, _place)) ->
-          print (draw_a x) (draw_a y)
-      | Job.Base None ->
-          print none ""
-      | Job.Base (Some (d, _place)) ->
-          draw_d d
+  (*
+    a -> 'a Merge.t
+    b -> New_job.t Data_list.t
+    c -> weight
+    d -> 'd Base.t
+    e -> 'a (final proof)
+    fa, fb are to update the nodes with new jobs and mark old jobs to "Done"*)
+  let rec update_split : type a b c d e.
```

Comment:
Consider giving the type variables better names to make the code clearer:

`type a_merge new_job_list weight base final_proof` or something, and then `f_merge` and `f_jobs` or something